File: 0017 - Letter Combinations Of A Phone Number - Medium.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        map = { 
        "2": "abc", 
        "3": "def", 
        "4": "ghi", 
        "5": "jkl",
        "6": "mno", 
        "7": "pqrs", 
        "8": "tuv", 
        "9": "wxyz"
        }

        if digits == "":
            return []

        letters = list(map[digits[0]])

        for digit in digits[1:]:
            letters = [old+new for old in letters for new in list(map[digit])]

        return letters

File: 0017 - Letter Combinations Of A Phone Number - Medium.java
// recursion approach

class Solution {
    public List<String> letterCombinations(String digits) {
        if (digits.isEmpty())
            return new ArrayList<>();

        return pad("", digits);
    }

    static ArrayList<String> pad(String p, String up) {
        if (up.isEmpty()) {
            ArrayList<String> list = new ArrayList<>();
            list.add(p);
            return list;
        }

        int digit = up.charAt(0) - '0';

        String[] map = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

        ArrayList<String> list = new ArrayList<>();
        if (digit < 2 || digit > 9) {
            return list;
        }
        String letters = map[digit];
        for (char ch : letters.toCharArray()) {
            list.addAll(pad(p + ch, up.substring(1)));
        }
        return list;
    }
}
